pkg/sql/descriptor.go

```diff
@@ -326,13 +331,45 @@ func GetAllDescriptors(
 			}
 			descs = append(descs, table)
 		case *sqlbase.Descriptor_Database:
+			dbDesc := desc.GetDatabase()
+			dbDescs[dbDesc.ID] = dbDesc
 			descs = append(descs, desc.GetDatabase())
 		case *sqlbase.Descriptor_Type:
-			descs = append(descs, desc.GetType())
+			typDesc := desc.GetType()
+			typDescs[typDesc.ID] = typDesc
+			descs = append(descs, typDesc)
 		default:
 			return nil, errors.AssertionFailedf("Descriptor.Union has unexpected type %T", t)
 		}
 	}
+
+	// If we found any type descriptors, that means that some of the tables we
+	// scanned might have types that need hydrating.
+	if len(typDescs) > 0 {
+		// Since we just scanned all the descriptors, we already have everything
+		// we need to hydrate our types. Set up an accessor for the type hydration
+		// method to look into the scanned set of descriptors.
+		typeLookup := func(id sqlbase.ID) (*tree.TypeName, *TypeDescriptor, error) {
+			typDesc := typDescs[id]
+			dbDesc := dbDescs[typDesc.ParentID]
+			schemaName, err := schema.ResolveNameByID(ctx, txn, codec, dbDesc.ID, typDesc.ParentSchemaID)
+			if err != nil {
```

Comment:
I think that it's worth being more defensive here. If we somehow got a corrupted table descriptor that references an enum that no longer exists, you'd be unable to do anything that required GetAllDescriptors, which is called all over the place. Would it be possible to log an error, but still return a descriptor that's just not hydrated, and add errors when you actually try to use an unhydrated descriptor?

Reply:
Hmm, that's a great point. Logging here makes sense. I'll see what can be done about getting errors from usages of the unhydrated descriptor (see #49404), but it is tricky to plumb errors out of some of these methods. The most important thing for these virtual tables is that the type names in the descriptors are resolved so that they display correctly. Without it, they just display with the unqualified name.